go/test/endtoend/shardedrecovery/sharded_recovery_test.go

```diff
@@ -0,0 +1,630 @@
+	assert.Nil(t, err)
+}
+
+func verifyQueriesUsingVtgate(t *testing.T, session *vtgateconn.VTGateSession, query string, value string) {
+	qr, err := session.Execute(context.Background(), query, nil)
+	assert.Nil(t, err)
+	assert.Equal(t, value, fmt.Sprintf("%v", qr.Rows[0][0]))
+}
+
+func executeQueriesUsingVtgate(t *testing.T, session *vtgateconn.VTGateSession, query string) {
+	_, err := session.Execute(context.Background(), query, nil)
+	assert.Nil(t, err)
+}
+
+func initializeCluster() (int, error) {
+
+	localCluster = cluster.NewCluster(cell, hostname)
+
+	// Start topo server
+	err := localCluster.StartTopo()
+	if err != nil {
+		return 1, err
+	}
+	// Start keyspace
+	keyspace := &cluster.Keyspace{
+		Name: keyspaceName,
+	}
+	localCluster.Keyspaces = append(localCluster.Keyspaces, *keyspace)
+
+	shard := cluster.Shard{
+		Name: shardName,
+	}
+	shard0 := cluster.Shard{
+		Name: shard0Name,
+	}
+	shard1 := cluster.Shard{
+		Name: shard1Name,
+	}
+
+	var mysqlProcs []*exec.Cmd
+	for _, eachShard := range []cluster.Shard{shard, shard0, shard1} {
+		tabletsRequired := 3
+		if eachShard.Name == shardName {
+			tabletsRequired = 5
+		}
+		for i := 0; i < tabletsRequired; i++ {
+			tabletType := "replica"
+			if i == 0 {
+				tabletType = "master"
+			}
+			if i == 2 {
+				tabletType = "rdonly"
+			}
+			tablet := localCluster.GetVttabletInstance(tabletType, 0, cell)
+			tablet.VttabletProcess = localCluster.GetVtprocessInstanceFromVttablet(tablet, eachShard.Name, keyspaceName)
+			tablet.VttabletProcess.ExtraArgs = commonTabletArg
+			tablet.VttabletProcess.SupportsBackup = true
+			tablet.VttabletProcess.EnableSemiSync = true
+			tablet.MysqlctlProcess = *cluster.MysqlCtlProcessInstance(tablet.TabletUID, tablet.MySQLPort, localCluster.TmpDirectory)
+			if proc, err := tablet.MysqlctlProcess.StartProcess(); err != nil {
+				return 1, err
+			} else {
+				mysqlProcs = append(mysqlProcs, proc)
+			}
+			eachShard.Vttablets = append(eachShard.Vttablets, tablet)
+		}
+		localCluster.Keyspaces[0].Shards = append(localCluster.Keyspaces[0].Shards, eachShard)
+	}
+
+	for _, proc := range mysqlProcs {
+		var stderr bytes.Buffer
+		proc.Stderr = &stderr
+		if err := proc.Wait(); err != nil {
+			return 1, err
+		}
+	}
+
+	master = localCluster.Keyspaces[0].Shards[0].Vttablets[0]
+	replica1 = localCluster.Keyspaces[0].Shards[0].Vttablets[1]
+	rdOnly = localCluster.Keyspaces[0].Shards[0].Vttablets[2]
+	replica2 = localCluster.Keyspaces[0].Shards[0].Vttablets[3]
+	replica3 = localCluster.Keyspaces[0].Shards[0].Vttablets[4]
+	shard0Master = localCluster.Keyspaces[0].Shards[1].Vttablets[0]
+	shard0Replica = localCluster.Keyspaces[0].Shards[1].Vttablets[1]
+	shard0RdOnly = localCluster.Keyspaces[0].Shards[1].Vttablets[2]
+	shard1Master = localCluster.Keyspaces[0].Shards[2].Vttablets[0]
+	shard1Replica = localCluster.Keyspaces[0].Shards[2].Vttablets[1]
+	shard1RdOnly = localCluster.Keyspaces[0].Shards[2].Vttablets[2]
```

Comment:
Maybe a better way to create these objects first and then add them into shard(s) and then loop those shards to create mysql processes. Check resharding_base for example.

Reply:
done